prow/cmd/deck/main.go

```diff
@@ -408,6 +431,171 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+// A valid job GCS Job view url takes the form:
+//
+// /view/gcs/<bucketname>/<jobprefix>
+//
+// Example:
+// - /view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/121
+func handleRequestGCSJobViews(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		setHeadersNoCaching(w)
+		start := time.Now()
+		srcData := strings.TrimPrefix(r.URL.Path, "/view/gcs/")
+		src := fmt.Sprintf("gs://%s", srcData)
+		fmt.Fprint(w, renderSpyglass(sg, ca, src))
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return GCS view page: ", elapsed)
+	}
+}
+
+// handleRequestJobViews handles requests to get all available artifact views for a given job via
+// a general src parameter, which can contain any string used to obtain job artifacts
+// A valid general job view url takes the form:
+//
+// /view/?src=<URI-encoded-source-string>
+//
+// Example:
+// - /view/?src=gs:%2F%2Fkubernetes-jenkins%2Fpr-logs%2Fpull%2Fkubeflow_kubeflow%2F1195%2Fkubeflow-presubmit%2F2558
+func handleRequestJobViews(sg *spyglass.Spyglass, ca *config.Agent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		setHeadersNoCaching(w)
+		start := time.Now()
+		src := r.URL.Query().Get("src")
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		fmt.Fprint(w, renderSpyglass(sg, ca, src))
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return view page: ", elapsed)
+	}
+}
+
+// renderSpyglass returns a pre-rendered Spyglass page from the given source string
+func renderSpyglass(sg *spyglass.Spyglass, ca *config.Agent, src string) string {
+	artifactNames, err := sg.ListArtifacts(src)
+	if err != nil {
+		logrus.WithError(err).Error("Error listing artifacts")
+	}
+
+	viewerCache := map[string][]string{}
+	viewersRegistry := ca.Config().Deck.Spyglass.Viewers
+
+	for re, viewerNames := range viewersRegistry {
+		matches := []string{}
+		r, err := regexp.Compile(re)
+		if err != nil {
+			logrus.WithError(err).Error("Regexp failed to compile.")
+			continue
+		}
+		for _, a := range artifactNames {
+			if r.MatchString(a) {
+				matches = append(matches, a)
+			}
+		}
+		for _, vName := range viewerNames {
+			viewerCache[vName] = matches
+		}
+	}
+
+	viewsStart := time.Now()
+	lenses := sg.Views(viewerCache)
+	viewsElapsed := time.Since(viewsStart)
+	logrus.Info("Got views in ", viewsElapsed)
+
+	var viewBuf bytes.Buffer
+	type ViewsTemplate struct {
+		Views       []spyglass.Lens
+		Source      string
+		ViewerCache map[string][]string
+	}
+	vTmpl := ViewsTemplate{
+		Views:       lenses,
+		Source:      src,
+		ViewerCache: viewerCache,
+	}
+	t := template.Must(template.ParseFiles("/static/template/spyglass-template.html"))
+	tErr := t.Execute(&viewBuf, vTmpl)
+	if tErr != nil {
+		logrus.WithError(tErr).Error("Error rendering template.")
```

Comment:
seems like we should return something special in this case

Reply:
yes this seems like a good place for a 500